File: 15-2.rb
#!/usr/bin/env ruby

def HASH(string)
  value = 0
  string.codepoints.each do |c|
    value = ((value + c) * 17) % 256
  end
  value
end

class Lens
  attr_reader :focal_length, :label

  def initialize(focal_length, label)
    @focal_length = focal_length.to_i
    @label = label
  end

  def inspect
    to_s
  end

  def to_s
    "<#{self.class}: #{@label} #{@focal_length}>"
  end
end

class Box
  def initialize(index)
    @index = index.to_i
    @lenses = []
  end

  def empty?
    @lenses.empty?
  end

  def power
    box_power = 0
    @lenses.each_with_index do |l, i|
      box_power += (@index + 1) * (i + 1) * l.focal_length
    end
    box_power
  end

  def insert(focal_length, label)
    if old_index = @lenses.index { |l| l.label == label }
      @lenses.insert(old_index, Lens.new(focal_length, label))
      @lenses.delete_at(old_index + 1)
    else
      @lenses << Lens.new(focal_length, label)
    end
  end

  def remove(label)
    @lenses.delete_if { |l| l.label == label }
  end

  def inspect
    to_s
  end

  def to_s
    s = "Box #{@index}: "
    s += @lenses.map(&:to_s).join ' '
  end
end

input = File.read('15.input').lines.map(&:strip)[0].split ','

boxes = []
256.times do |i|
  boxes << Box.new(i)
end
input.each do |instruction|
  if instruction.include? '='
    label, focal_length = instruction.split '='
    box_num = HASH label
    boxes[box_num].insert(focal_length, label)
  elsif instruction.end_with? '-'
    label = instruction.split('-')[0]
    box_num = HASH label
    boxes[box_num].remove(label)
  else
    abort "Parse error: #{instruction}"
  end
end

print boxes.map(&:power).sum, "\n"

File: 15-2.py
#!/usr/bin/env python

def HASH(string):
    value = 0
    for c in string:
        value = ((value + ord(c)) * 17) % 256
    return value


class Lens:
    def __init__(self, focal_length, label):
        self.focal_length = int(focal_length)
        self.label = label


class Box:
    def __init__(self, index):
        self.index = int(index)
        self.lenses = []

    def is_empty(self):
        return len(self.lenses) == 0

    def power(self):
        box_power = 0
        for i in range(len(self.lenses)):
            box_power += (self.index + 1) * (i + 1) * self.lenses[i].focal_length
        return box_power

    def insert(self, focal_length, label):
        old_index = -1
        for i in range(len(self.lenses)):
            if self.lenses[i].label == label:
                old_index = i
                break
        if old_index != -1:
            self.lenses.insert(old_index, Lens(focal_length, label))
            del self.lenses[old_index + 1]
        else:
            self.lenses.append(Lens(focal_length, label))

    def remove(self, label):
        found = -1
        for i in range(len(self.lenses)):
            if self.lenses[i].label == label:
                found = i
                break
        if found != -1:
            del self.lenses[i]


input = open('15.input').readlines()[0].strip().split(",")

boxes = []
for i in range(256):
    boxes.append(Box(i))

for instruction in input:
    if "=" in instruction:
        label, focal_length = instruction.split("=")
        box_num = HASH(label)
        boxes[box_num].insert(focal_length, label)
    elif "-" in instruction:
        label = instruction.split("-")[0]
        box_num = HASH(label)
        boxes[box_num].remove(label)
    else:
        exit(1)

print(sum([b.power() for b in boxes]))
